envoy/matcher/matcher.h

```diff
@@ -151,6 +153,95 @@ template <class DataType> class MatchTree {
 
 template <class DataType> using MatchTreeSharedPtr = std::shared_ptr<MatchTree<DataType>>;
 
+// Generic container for an input value.
+class InputValue {
+public:
+  InputValue() {}
+  InputValue(absl::string_view data) : data_(absl::in_place_type<std::string>, data) {}
+  InputValue(int64_t data) : data_(data) {}
+
+  enum class Kind {
+    // Null value (use when value is not present).
+    Null,
+    // String value (possibly invalid UTF-8).
+    String,
+    // 64-bit signed integer value.
+    Int,
+    // Dynamic list value (possibly heterogenous).
+    List
+  };
+
+  Kind kind() const {
+    switch (data_.index()) {
+    case 1:
+      return Kind::String;
+    case 2:
+      return Kind::Int;
+    case 3:
+      return Kind::List;
+    default:
+      return Kind::Null;
+    }
+  }
+
+  bool isNull() const { return kind() == Kind::Null; }
+
+  /**
+   * Access the stored string value or throw an exception if the kind does not match.
```

Comment:
Throwing an exception possibly on the data plan path seems icky, is there a way to avoid that?